src/engine/thread.js

```diff
@@ -254,17 +270,20 @@ class Thread {
      * @param {string} blockId Block ID to push to stack.
      */
     reuseStackForNextBlock (blockId) {
-        this.stack[this.stack.length - 1] = blockId;
-        this.stackFrames[this.stackFrames.length - 1].reuse();
+        this.pointer = blockId;
+        if (this.stackFrame.needReset) this.stackFrame.reuse();
     }
 
     /**
      * Pop last block on the stack and its stack frame.
      * @return {string} Block ID popped from the stack.
      */
     popStack () {
-        _StackFrame.release(this.stackFrames.pop());
-        return this.stack.pop();
+        const popped = this.pointer;
```

Comment:
`popped` is a weird name for this given the code. Maybe `lastPointer` or `lastBlock` would be better?
I think we should also update the comment on this function so that it doesn't necessarily have to reference stack functionality directly (even though the function name itself does). Maybe something like:

'Remove the last block from the stack and its stack frame. Update the current instruction pointer.' or something...